File: sdks/java/src/main/java/com/blok/blok/nodes/HelloWorldNode.java
package com.blok.blok.nodes;

import com.blok.blok.node.NodeHandler;
import com.blok.blok.types.Context;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

/**
 * Simple example node that greets the user.
 * <p>
 * Input (request body):
 * <ul>
 *   <li>name (string, optional) - Name to greet (default: "World")</li>
 * </ul>
 * Config:
 * <ul>
 *   <li>prefix (string, optional) - Greeting prefix (default: "Hello")</li>
 * </ul>
 * Output:
 * <ul>
 *   <li>message (string) - The greeting message</li>
 *   <li>timestamp (string) - ISO 8601 timestamp</li>
 *   <li>language (string) - "java"</li>
 * </ul>
 */
public class HelloWorldNode implements NodeHandler {

    @Override
    public Object execute(Context ctx, Map<String, Object> config) {
        // Get name from request body, default to "World"
        String name = "World";
        Map<String, Object> body = ctx.getRequest().bodyMap();
        if (body != null) {
            Object nameVal = body.get("name");
            if (nameVal instanceof String s && !s.isEmpty()) {
                name = s;
            }
        }

        // Get prefix from config, default to "Hello"
        String prefix = "Hello";
        if (config != null) {
            Object prefixVal = config.get("prefix");
            if (prefixVal instanceof String s && !s.isEmpty()) {
                prefix = s;
            }
        }

        String message = prefix + ", " + name + "!";

        // Store greeting in context vars for downstream nodes
        ctx.setVar("greeting", message);

        Map<String, Object> result = new HashMap<>();
        result.put("message", message);
        result.put("timestamp", Instant.now().toString());
        result.put("language", "java");
        return result;
    }
}

File: sdks/python3/examples/hello_world_node.py
from __future__ import annotations
import logging
from datetime import datetime, timezone
from typing import Any, Dict

from blok.node.node_handler import NodeHandler
from blok.types.context import Context

# Logger that flows into the runner's RunTracker when BLOK_STREAM_LOGS=true.
# Reusing the canonical name documented in blok.server.grpc_server.NODE_LOGGER_NAME
# so handler authors see a consistent convention across examples.
log = logging.getLogger("blok.node")


class HelloWorldNode(NodeHandler):
    """Greets the user with a configurable prefix.

    Config:
        prefix (str, optional): Greeting prefix (default: "Hello")

    Request body:
        name (str, optional): Name to greet (default: "World")

    Output:
        {"message": "Hello, World!", "timestamp": "...", "language": "python3"}
    """

    def execute(self, ctx: Context, config: Dict[str, Any]) -> Any:
        name = ctx.request.body_str("name") or "World"
        prefix = config.get("prefix", "Hello")
        message = f"{prefix}, {name}!"

        # Emitted on the `blok.node` logger so the runner's streaming path
        # forwards it as a LogLine event when BLOK_STREAM_LOGS=true. Harmless
        # in HTTP/unary mode — the logger is just a normal Python logger.
        log.info("greeting %s with prefix %r", name, prefix)

        ctx.set_var("greeting", message)

        return {
            "message": message,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "language": "python3",
        }
